[fx] GraphModule copy top level attributes from root

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45182 [fx] GraphModule copy top level attributes from root**

Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23858632](https://our.internmc.facebook.com/intern/diff/D23858632)

jerryzh168 committed:
[fx] GraphModule copy top level attributes from root

Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx] GraphModule copy top level attributes from root"


Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -132,8 +132,7 @@ def __init__(self, root: Union[torch.nn.Module, Dict[str, Any]], graph: Graph):
         if isinstance(root, torch.nn.Module):
-            if hasattr(root, 'training'):
-                self.training = root.training
+            self.__dict__ = root.__dict__
```
I don't think we want this behavior in the normal constructor. It's probably better to just modify `__deepcopy__` rather than make this change

jerryzh168 committed:
Update on "[fx] GraphModule copy top level attributes from root"


Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23858632](https://our.internmc.facebook.com/intern/diff/D23858632)

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -194,7 +194,9 @@ def __reduce__(self):
         fake_mod.__dict__ = copy.deepcopy(self.__dict__)
-        return GraphModule(fake_mod, self.graph)
+        graph_module = GraphModule(fake_mod, self.graph)
+        graph_module.__dict__ = fake_mod.__dict__
```
Does this contain generated artifacts (`.code` or `.forward`)? If so, you should delete tehm

jerryzh168 replied:
it contains:
dict_keys(['training', '_parameters', '_buffers', '_non_persistent_buffers_set', '_backward_hooks', '_forward_hooks', '_forward_pre_hooks', '_state_dict_hooks', '_load_state_dict_pre_hooks', '_modules', '_graph', 'code', 'attr'])

will delete _modules, _graph and code.

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -194,7 +194,9 @@ def __reduce__(self):
+        return graph_module
 
     def __copy__(self):
         return GraphModule(self, self.graph)
```
What about here too?

jerryzh168 committed:
Update on "[fx] GraphModule copy top level attributes from root"


Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23858632](https://our.internmc.facebook.com/intern/diff/D23858632)

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -194,10 +194,19 @@ def __reduce__(self):
+        # delete generated attributes before creating a new
+        # GraphModule
+        generated_attrs = ['code', '_graph', '_modules']
+        for attr in generated_attrs:
```
`generated_attrs` is only used once, so you can probably just put the list literal in the loop statement

jamesr66a commented on torch/fx/graph_module.py:
```diff
@@ -194,10 +194,19 @@ def __reduce__(self):
     def __copy__(self):
-        return GraphModule(self, self.graph)
+        graph_module = GraphModule(self, self.graph)
+        graph_module.__dict__ = self.__dict__
```
Do we need to delete the generated attributes here as well?

jerryzh168 replied:
I don't think we can delete them, since it is used by the old module, but I skipped overwriting them

jerryzh168 committed:
Update on "[fx] GraphModule copy top level attributes from root"


Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23858632](https://our.internmc.facebook.com/intern/diff/D23858632)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[fx] GraphModule copy top level attributes from root"


Summary:
Previously only the attributes that's used in the graph is copied in the constructor
of GraphModule, this PR adds the support for copying all attributes by overriding `__dict__`

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23858632](https://our.internmc.facebook.com/intern/diff/D23858632)

[ghstack-poisoned]

zdevito commented on torch/fx/graph_module.py:
```diff
@@ -194,10 +194,20 @@ def __reduce__(self):
-        return GraphModule(fake_mod, self.graph)
+        graph_module = GraphModule(fake_mod, self.graph)
+        # skip overwriting generated attributes
+        for attr in ['code', '_graph', '_modules']:
```
These fields are not complete. For instance `_parameters` and `_buffers` are not included and copying them might break the nn.Module similar to `_modules`.  Any attempt to list parameters here would be very fragile because changes to `nn.Module` would affect what this list needs to be with no reliable way to inform an editor of nn.Module that this list also needs to be modified.

jerryzh168 replied:
actually I'm not sure why I included `_modules` we probably don't need that. This is just for skipping attributes generated when creating a GraphModule.

jerryzh168 commented:
fixed in the original PR: https://github.com/pytorch/pytorch/pull/44074
